syntax = "proto2";

import "nanopb.proto";

// Regen command for python:
// protoc --proto_path=arduino/proto --python_out=protoc arduino/proto/motor_command.proto

message MotorInitProto {
  // Required?!?! Why?
  // It turns out that nanopb ignores optional fields, hence the deviation from
  // Google norms.
  required int32 address = 1;
  required int32 enable_pin = 2;
  required int32 dir_pin = 3;
  required int32 step_pin = 4;
}

message MotorMoveProto {
  // Speeds are between 0.0 and 1.0.
  required float max_speed = 1;
  // We start at this min speed and will try to decelerate back to it.
  required float min_speed = 2;
  required bool disable_after_moving = 3;
  // If set, used instead of steps.
  required int32 absolute_steps = 4;
  required float acceleration = 5;
}

message MotorConfigProto {
  required int32 address = 1;
  required bool zero = 5;
  required int32 min_steps = 6;
  required int32 max_steps = 7;
}

message MotorMoveAllProto {
  repeated MotorMoveProto motors = 1 [(nanopb).max_count = 6, (nanopb).fixed_count = true];
}

message MotorTareProto {
  required int32 address = 1;
  required int32 tare_to_steps = 2;
}

message MotorReportProto {
  required int32 current_absolute_steps = 1;
  required float acceleration = 2;
  required float step_progress = 3;
  required float step_speed = 4;
}

message AllMotorReportProto {
  repeated MotorReportProto motors = 1 [(nanopb).max_count = 6, (nanopb).fixed_count = true];
}

// Indicates pins to read from. Does not disable unset pins; once set to input, that's what a pin does. 
// pullup supercedes input if both are set.
message IOReadRequestProto {
  required int64 input_pin_bitmap = 1;
  required int64 pullup_pin_bitmap = 2;
}

message IOReadProto {
  required int64 pin_states_bitmap = 1;
}

// If pin_to_watch reaches pin_state_to_match, the given motor is tared to tare_to_steps.
message MotorTareIfProto {
  required int32 address = 1;
  required int32 tare_rule_index = 2;
  required int32 tare_to_steps = 3;
  required uint32 pin_to_watch = 4;
  required bool pin_state_to_match = 5;
}

